### ISSUE TITLE
Can't add/upload a file via WebDav on Windows 10

### ISSUE THREAD

**Issue by stevensam** [civetweb/civetweb#1040]
Hi,

I'm trying to use WebDav with Civetweb 64-bit (precompiled V1.15) on Windows 10 64-bit. It appears to work fine for downloading a file via the Explorer window from Civitweb running on localhost (i.e., \\localhost@8080\DavWWWRoot). 

However when trying to upload a file or folder, I'm getting a "Replace or Skip files" dialog window, and when I click "Replace the file in the destination" (which appears even though the file didn't exist there before ), I then get a dialog window with "1 Interrupted Action  -- Can't read from source file or disk" and button for:  'Try Again | Skip | Cancel'  Pressing 'Skip' appears to create an empty file.  

Oddly, I can delete files using the Explorer window just fine.

I'm trying to upload simple text files without any special permissions. I've tried running Civetweb under Admin permissions, and specifying a "put_delete_auth_file". I wasn't sure what the format of the "put_delete_auth_file" should be or how to create it however (I couldn't find any documentation on that), so I just used the file generated with the "Modify passwords file" function thinking that might be the same format...

Can someone explain how to setup WebDav so that files can be uploaded to a directory on Windows 10? 

Thank you!

**Comment by bel2125** [comment]
You already found the information how to create an auth file.
I just tested WebDav using WinSCP and this worked ... you tested this in your update as well, so WebDav is working in general.

The problem with Windows 10 (and Windows 8 and 7 before): it seems to require manually changing a lot of registry keys. I got the error message "davwwwroot is not accessible" ... googling this shows a lot of SharePoint Q&As, adding "-sharepoint" to the search query shows some pages that explain heavy Windows registry fiddling.
I did not risk to try this on my main PC now - probably I can test it later on a test PC or VM.
For webdav using https you will need a certificate that is trusted by Windows - so you need to install this certificate as well.

**Comment by bel2125** [comment]
With the following settings I got Windows 10 explorer "mostly" working as WebDav client: 

First set the `authentication_domain` to your server host name, e.g. to "localhost". You have to use a name that is in available from the DNS for the Windows 10 client PC.
Set the `listening_port` to 80 (using http for now).
Create a .htpasswd file with user name "user", password "pass" and domain "localhost". Of course you can choose user name and password, but for the domain you must use the same name as in the `authentication_domain` setting.
Then set the path to this file for `global_auth_file` as well as for `put_delete_auth_file`.
I followed [these instructions](https://uvacollab.screenstepslive.com/s/help/m/sitetools/l/611578-how-do-i-set-up-webdav-for-windows-10-or-8) and added a network location.

Works without problems:
- Then I can open/read files without problems.
- I can copy files from the server using drag & drop.
- I can also drop entire folder structures using drag & drop.
- I can also delete files without problems.
- I can delete folders recursively.

Works but shows some warnings:
- Right click "new folder" shows a "Permission error", but the folder will be created.
- Adding a file by drag&drop will copy this file. But after it is copied, there is a dialog *the destination already has a file named "file.ext"*: Replace/Skip/Compare. You can click "Skip", the file is already there.
- You can also drag&drop entire folders onto the server. You will see the same "already exists" issue.

Does not work (but I'm not sure if this use case makes sense at all):
- Right click "create new file" shows a "Error 0x8000FFFF: Catastrophic failure" ... but nothing happens. If you ever create a file using this right click + new file, you have to do this on a local disk and then upload the final document.

On the PC I tested this, I did not have to modify the Windows registry. But I know this PC has used Web folders (SharePoint) before, so probably some admin already changed something here.

**Comment by bel2125** [comment]
I just realized, the files transferred using Windows 10 explorer will have 0 byte size if it is done like this, while it is OK for other WebDav clients. I think this needs some more time for a closer investigation.

**Comment by stevensam** [comment]
> I just realized, the files transferred using Windows 10 explorer will have 0 byte size if it is done like this, while it is OK for other WebDav clients. I think this needs some more time for a closer investigation.

Indeed, this is the crux of the problem. I also tested it using WinSCP and it does work with that, however it doesn't with other WebDav clients like CyberDuck. 

Thank you for your thorough testing and instructions though! I had researched and tried many of the same things, too such as changing the registry keys and setting the auth files (which I did get working). Nothing seemed to resolve the issue though.

I also tried using WebDav with Apache without SSL and didn't experience the same issue with Explorer. 

Perhaps turning on the debug log will indicate what's failing during `Put`, or the Windows Event Viewer log will contain a clue.

**Comment by bel2125** [comment]
The Windows 10 Explorer uses the "LOCK" method, that is not implemented in CivetWeb.
If you only want to upload/download files like WinSCP does, you don't need this method.
WinSCP uses "PUT" to upload a file, and this works.
However, Windows 10 Explorer seems to start with an empty "PUT" followed by a "LOCK" (for exclusive access to the newly generated file). This is currently not supported.
I can check what it would mean to support this, maybe it is simple to add, but right now it cannot work like this.

**Comment by stevensam** [comment]
> However, Windows 10 Explorer seems to start with an empty "PUT" followed by a "LOCK" (for exclusive access to the newly generated file). This is currently not supported.

That sounds right - I believe I saw mention of that in MSDN somewhere while researching this. That's great that you figured it out, so I can finally stop testing and trying to figure out workarounds :-)  (Using SSL or connecting over a network didn't work either of course)

> I can check what it would mean to support this, maybe it is simple to add, but right now it cannot work like this.

That would be awesome! Having WebDav in CivetWeb be fully compatible with the built-in webclient in Windows (i.e., Explorer) would be super helpful.

Thanks so much for the quick replies and all of the testing & research!

**Comment by stevensam** [comment]
By the way, I gather that CivetWeb is being used as a WebDav server for _Orthanc_ based on [this ](https://github.com/civetweb/civetweb/pull/939) thread, and Explorer in Windows 10 might be working with that based on their documentation [here](https://book.orthanc-server.com/users/webdav.html)  (?)

 If so, perhaps they might have a patch? It seems like it would be important to get that working for them as well if not.

**Comment by bel2125** [comment]
This patch is already merged: https://github.com/civetweb/civetweb/pull/939 
The LOCK/UNLOCK/PROPPATCH methods are probably managed by an additional script or callback.

Anyway, I will need some time to have a closer look at the Windows 10 client.

**Comment by stevensam** [comment]
Sure, thanks!

**Comment by bel2125** [comment]
Unfortunately there is no usable documentation on what Windows 10 Explorer really expects from a WebDAV server as a minimum. However, I managed to get it running on my test PC by adding the "LOCK"/"UNLOCK" and "PROPPATCH" method. These new implementations are effectively doing nothing except sending some response to Windows Explorer. In particular, there is no "Property storage" implemented. It seems OK for Windows Explorer if every "PROPPATCH" request is answered with a "403 Forbidden", as long as this "403" is packed into a verbose XML response.

The current "master" source code can be used for testing this feature. If all tests succeed, this can be part of the next release. But I would need some help with testing on different Windows systems - "it works on my PC" won't help anyone else. Maybe different Windows versions work differently here, or maybe different ways to Windows Explorer require additional or modified implementations.

**Comment by stevensam** [comment]
Great - sounds promising! 

> The current "master" source code can be used for testing this feature. If all tests succeed, this can be part of the next release. But I would need some help with testing on different Windows systems...

No problem, I have various Windows computers (not just VM's) to test this on for Windows XP through Windows 11, for both 32-bit and 64-bit (up until Windows 10).

If added to the master, then I'll compile and test it using a matrix of the above OS/architecture combinations, and will report back the results.  

Just comment back when it's available, and if there's anything specific I should do or test - other than to check all the typical WebDav functions using Explorer.

Thanks again!

**Comment by bel2125** [comment]
Yes, the functionality should be available with the "WebDAV for Windows Explorer (Step 3/?)" commit. The code still needs some cleanup and there is no documentation yet, and it probably won't work if two civetweb instances are started within the same C/C++ program - but I think the code is ready for a test.

**Comment by stevensam** [comment]
Got it, thanks. It might take a few days for me to fire up all the PC's, update their OS's, and then compile and test each function. I'll let you know when I'm done. 

If there's anything else you'd like me to test while I'm at it, just comment back. I'm happy to help out with testing future releases too.

**Comment by bel2125** [comment]
The improved version should also support renaming files and using non-ASCII characters.
The only remaining issue is a strange behavior when a % character is used in the file names, probably Windows-Explorer is trying to url-decode them twice.
